src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -172,135 +185,144 @@ public bool ParseHeaders(TRequestHandler handler, ref SequenceReader<byte> reade
-                if (length > 0)
-                {
-                    continue;
-                }
+                    if (foundCrlf)
+                    {
+                        eolSize = 2;
 
-                // We moved the reader to look ahead 2 bytes so rewind the reader
-                if (readAhead > 0)
-                {
-                    reader.Rewind(readAhead);
+                        // Advance past the LF too
+                        reader.Advance(1);
+
+                        // Empty line?
+                        if (crIndex == 0)
+                        {
+                            handler.OnHeadersComplete(endStream: false);
+                            return true;
+                        }
+                    }
                 }
+                else
+                {
+                    // We got an LF with no CR before it.
+                    var lfIndex = lfOrCrIndex;
+                    if (!_allowLineFeedTerminator)
+                    {
+                        RejectRequestHeader(AppendEndOfLine(span[..lfIndex], lineFeedOnly: true));
+                    }
+
+                    // Consume the header including the LF
+                    reader.Advance(lfIndex + 1);
 
-                length = ParseMultiSpanHeader(handler, ref reader);
+                    span = span.Slice(0, lfIndex);
+                    if (span.Length == 0)
+                    {
+                        handler.OnHeadersComplete(endStream: false);
+                        return true;
+                    }
+                }
+            }
+            else
+            {
+                // No CR or LF. Is this a multi-span header?
+                int length = ParseMultiSpanHeader(handler, ref reader);
                 if (length < 0)
                 {
-                    // Not there
+                    // Not multi-line, just bad.
                     return false;
                 }
 
+                // This was a multi-line header. Advance the reader.
                 reader.Advance(length);
-                // As we crossed spans set the current span to default
-                // so we move to the next span on the next iteration
-                span = default;
+
+                continue;
+            }
+
+            // We got to a point where we believe we have a header.
+            if (!TryTakeSingleHeader(handler, span))
+            {
+                // Sequence needs to be CRLF and not contain an inner CR not part of terminator.
+                // Not parsable as a valid name:value header pair.
+                RejectRequestHeader(AppendEndOfLine(span, lineFeedOnly: eolSize == 1));
             }
         }
 
         return false;
     }
 
+    private static byte[] AppendEndOfLine(ReadOnlySpan<byte> span, bool lineFeedOnly)
+    {
+        var array = new byte[span.Length + (lineFeedOnly ? 1 : 2)];
+
+        span.CopyTo(array);
+        array[^1] = ByteLF;
+
+        if (!lineFeedOnly)
```

Comment:
This method could be written as
```c#
private static byte[] AppendEndOfLine(ReadOnlySpan<byte> span, bool lineFeedOnly)
{
    Debug.Assert(!span.IsEmpty);

    var array = new byte[span.Length + (lineFeedOnly ? 1 : 2)];

    Debug.Assert(array.Length >= 2);

    array[^2] = ByteCR;
    array[^1] = ByteLF;

    span.CopyTo(array);
    return array;
}
```
So just always write the CR/LF sequence, then in CopyTo it gets overwritten.
This saves the branch from `if (!lineFeedOnly)`. Note: I expecte PGO will emit similar code (+ the fallback code-path if the assumption isn't met), but this saves total IL and machine code.

---

<details>
  <summary>going "Unsafe"</summary>

One could also go "Unsafe" to remove the bound-checks from `array[^2]` (and `^1`):
```c#
private static byte[] AppendEndOfLine(ReadOnlySpan<byte> span, bool lineFeedOnly)
{
    Debug.Assert(!span.IsEmpty);

    var array = new byte[span.Length + (lineFeedOnly ? 1 : 2)];

    Debug.Assert(array.Length >= 2);

    ref byte ra = ref MemoryMarshal.GetArrayDataReference(array);
    ra = ref Unsafe.Add(ref ra, (uint)(array.Length - 2));
    Unsafe.As<byte, short>(ref ra) = BitConverter.IsLittleEndian
        ? (short)(ByteCR | ByteLF << 8)
        : (short)(ByteLF | ByteCR << 8);

    span.CopyTo(array);
    return array;
}
```
</details>

In a micro-benchmark it looks like:
```
| Method | LineFeedOnly |     Mean |     Error |    StdDev | Ratio | RatioSD |
|------- |------------- |---------:|----------:|----------:|------:|--------:|
|      A |        False | 9.698 ns | 0.2782 ns | 0.2977 ns |  1.00 |    0.00 |
|      B |        False | 9.473 ns | 0.1528 ns | 0.1355 ns |  0.98 |    0.03 |
|      C |        False | 9.185 ns | 0.2027 ns | 0.1896 ns |  0.95 |    0.03 |
|        |              |          |           |           |       |         |
|      A |         True | 9.795 ns | 0.1742 ns | 0.1544 ns |  1.00 |    0.00 |
|      B |         True | 9.326 ns | 0.2529 ns | 0.3105 ns |  0.96 |    0.04 |
|      C |         True | 9.699 ns | 0.2736 ns | 0.3461 ns |  0.99 |    0.04 |
```
A...method in this PR
B...suggestion in this comment
C...."Unsafe"-suggestion in this comment

I'd go with B, as it's still safe and nice to read.

Reply:
@sebastienros for my edification: didn't this show an improvement, as it got reverted (https://github.com/dotnet/aspnetcore/pull/43202/commits/6c6925941df71d66963bc4b7d73f80f75edb0ba3). I just like to understand if / how PGO, etc. comes into play here to learn from it.

Reply:
It was breaking the unit tests, so something is different from the previous implementation and at this point (RC1 is super close) I am not taking any chances.